### PR TITLE
Visitor should never panic.

### PR DESCRIPTION
**Pull request by pimeys**
Adresses (at least) these issues:

* https://github.com/prisma/prisma-client-js/issues/705
* https://github.com/prisma/quaint/issues/137
* https://github.com/prisma/prisma/issues/2561

### CODE REVIEW

File: src/connector/mysql/conversion.rs [hunk 1]
```diff
@@ -1,18 +1,77 @@
...
+                Value::Integer(i) => i.map(|i| my::Value::Int(i)),
+                Value::Real(f) => f.map(|f| my::Value::Double(f.to_f64().expect("Decimal is not a f64."))),
+                Value::Text(s) => s.clone().map(|s| my::Value::Bytes((&*s).as_bytes().to_vec())),
+                Value::Bytes(bytes) => bytes.clone().map(|bytes| my::Value::Bytes(bytes.into_owned())),
+                Value::Enum(s) => s.clone().map(|s| my::Value::Bytes((&*s).as_bytes().to_vec())),
+                Value::Boolean(b) => b.map(|b| my::Value::Int(b as i64)),
+                Value::Char(c) => c.map(|c| my::Value::Bytes(vec![c as u8])),
+                #[cfg(feature = "json-1")]
+                Value::Json(json) => json.clone().map(|json| {
+                    let s = serde_json::to_string(&json).expect("Cannot convert JSON to String.");
```
Comment (tomhoule) [inline src/connector/mysql/conversion.rs:31]: I don't know if this was overlooked or it's intended to still panic.

File: src/connector/mysql/conversion.rs [hunk 2]
```diff
@@ -1,18 +1,77 @@
...
-        my::Params::Empty
+        Ok(my::Params::Empty)
     } else {
-        my::Params::Positional(params.iter().map(|x| x.into()).collect::<Vec<my::Value>>())
+        let mut values = Vec::with_capacity(params.len());
+
+        for pv in params {
+            let res = match pv {
+                Value::Integer(i) => i.map(|i| my::Value::Int(i)),
+                Value::Real(f) => f.map(|f| my::Value::Double(f.to_f64().expect("Decimal is not a f64."))),
```
Comment (tomhoule) [inline src/connector/mysql/conversion.rs:23]: I don't know if this was overlooked or it's intended to still panic.

File: src/visitor/sqlite.rs [hunk 3]
```diff
@@ -50,7 +51,15 @@ impl<'a> Visitor<'a> for Sqlite<'a> {
             #[cfg(feature = "json-1")]
             Value::Json(j) => j.map(|j| self.write(format!("'{}'", serde_json::to_string(&j).unwrap()))),
```
Comment (tomhoule) [inline src/visitor/sqlite.rs:52]: not in the diff, but there is an unwrap here

File: src/visitor/mysql.rs [hunk 4]
```diff
@@ -66,7 +67,15 @@ impl<'a> Visitor<'a> for Mysql<'a> {
             #[cfg(feature = "json-1")]
             Value::Json(j) => j.map(|j| self.write(format!("CONVERT('{}', JSON)", serde_json::to_string(&j).unwrap()))),
```
Comment (tomhoule) [inline src/visitor/mysql.rs:68]: not in the diff, but there is an unwrap here

File: src/connector/sqlite/conversion.rs [hunk 5]
```diff
@@ -1,73 +1,131 @@
...
+                    c if c.is_bool() => Value::Boolean(None),
+                    c => match c.decl_type() {
+                        Some(n) => {
+                            let msg = format!("Value {} not supported", n);
+                            let kind = ErrorKind::conversion(msg);
+
+                            Err(Error::builder(kind).build())?
+                        }
+                        None => Value::Integer(None),
+                    },
```
Comment (tomhoule) [inline src/connector/sqlite/conversion.rs:125]: much better!

File: src/connector/type_identifier.rs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    fn is_datetime(&self) -> bool;
+    fn is_time(&self) -> bool;
+    fn is_date(&self) -> bool;
+    fn is_text(&self) -> bool;
+    fn is_bytes(&self) -> bool;
+    fn is_bool(&self) -> bool;
+    fn is_json(&self) -> bool;
+    fn is_enum(&self) -> bool;
+    fn is_null(&self) -> bool;
+}
```
Comment (tomhoule) [inline src/connector/type_identifier.rs:13]: :100:

Comment (pimeys) [reply src/connector/type_identifier.rs:13]: I tried to use these with pg, but it doesn't really fit there that well...

File: src/connector/sqlite/conversion.rs (hunk 5 above)
Comment (pimeys) [reply src/connector/sqlite/conversion.rs:125]: Yeah. Not perfect, but at least better...

**Pull request merged**